lms/djangoapps/course_api/blocks/transformers/milestones.py

```diff
@@ -35,39 +35,49 @@ def collect(cls, block_structure):
-        # this needs to change ^^^
-
-        def has_milestones_for_user(block_key):
+        def user_gated_from_block(block_key):
             """
-            Test whether the block is a proctored exam for the user in
-            question, or requires any other unfulfilled milestones.
+            Checks whether the user is gated from accessing this block, first via exam proctoring, then via a general
+            milestones check.
             """
-            if (
-                    block_key.block_type == 'sequential' and (
-                        block_structure.get_xblock_field(block_key, 'is_proctored_enabled') or
-                        block_structure.get_xblock_field(block_key, 'is_practice_exam')
-                    )
-            ):
-                # This section is an exam.  It should be excluded unless the
-                # user is not a verified student or has declined taking the exam.
-                user_exam_summary = get_attempt_status_summary(
-                    usage_info.user.id,
-                    unicode(block_key.course_key),
-                    unicode(block_key),
-                )
-                exam_is_proctored = user_exam_summary and \
-                    user_exam_summary['status'] != ProctoredExamStudentAttemptStatus.declined
+            return self.is_proctored_exam(block_key, usage_info, block_structure) or \
+                self.has_pending_milestones_for_user(block_key, usage_info)
 
-                # after checking for proctored, we check for all other milestones
-                # if we need additional fields bump up version number
-                milestones_exist_for_exam = bool(milestones_helpers.get_course_content_milestones(
-                    unicode(block_key.course_key),
-                    unicode(block_key),
-                    'requires',
-                    usage_info.user.id
-                ))
-                return exam_is_proctored or milestones_exist_for_exam
+        return [block_structure.create_removal_filter(user_gated_from_block)]
+
+    @staticmethod
+    def is_proctored_exam(block_key, usage_info, block_structure):
+        """
+        Test whether the block is a proctored exam for the user in
+        question.
+        """
+        if (
+                block_key.block_type == 'sequential' and (
+                    block_structure.get_xblock_field(block_key, 'is_proctored_enabled') or
+                    block_structure.get_xblock_field(block_key, 'is_practice_exam')
+                )
+        ):
+            # This section is an exam.  It should be excluded unless the
+            # user is not a verified student or has declined taking the exam.
+            user_exam_summary = get_attempt_status_summary(
+                usage_info.user.id,
+                unicode(block_key.course_key),
+                unicode(block_key),
+            )
+            return settings.FEATURES.get('ENABLE_PROCTORED_EXAMS', False) and user_exam_summary \
```

Comment:
Yeah, the setting was renamed in this commit from last September: https://github.com/edx/edx-platform/commit/9476898df0c82d2b598a9e9d9975c0301251517a#diff-82bda515879c96ab0641bfcbe58631c3, it should be `ENABLE_SPECIAL_EXAMS`.

That means that this transformer has always been short-circuiting its logic due to checking an undefined setting: https://github.com/edx/edx-platform/blob/master/lms/djangoapps/course_api/blocks/transformers/proctored_exam.py#L37